README.md
<!-- {{{1

    File        : README.md
    Maintainer  : Felix C. Stegerman <flx@obfusk.net>
    Date        : 2021-07-07

    Copyright   : Copyright (C) 2021  Felix C. Stegerman
    Version     : v0.2.3
    License     : AGPLv3+

}}}1 -->

[![GitHub Release](https://img.shields.io/github/release/obfusk/kanjidraw.svg?logo=github)](https://github.com/obfusk/kanjidraw/releases)
[![PyPI Version](https://img.shields.io/pypi/v/kanjidraw.svg)](https://pypi.python.org/pypi/kanjidraw)
[![Python Versions](https://img.shields.io/pypi/pyversions/kanjidraw.svg)](https://pypi.python.org/pypi/kanjidraw)
[![CI](https://github.com/obfusk/kanjidraw/workflows/CI/badge.svg)](https://github.com/obfusk/kanjidraw/actions?query=workflow%3ACI)
[![AGPLv3+](https://img.shields.io/badge/license-AGPLv3+-blue.svg)](https://www.gnu.org/licenses/agpl-3.0.html)
[![Sponsor](https://img.shields.io/badge/%E2%99%A5-support-violet.svg)](https://ko-fi.com/obfusk)

<a href="https://repology.org/project/kanjidraw/versions">
  <img src="https://repology.org/badge/vertical-allrepos/kanjidraw.svg?header="
    alt="Packaging status" align="right" />
</a>

<a href="https://repology.org/project/python:kanjidraw/versions">
  <img src="https://repology.org/badge/vertical-allrepos/python:kanjidraw.svg?header="
    alt="Packaging status" align="right" />
</a>

# kanjidraw

## handwritten kanji recognition

`kanjidraw` is a simple Python library + GUI for matching (the strokes
of a) handwritten kanji against its database.

You can use the GUI to draw and subsequently select a kanji from the
list of probable matches, which will then be copied to the clipboard.

The database is based on KanjiVG and the algorithms are based on the
[Kanji draw](https://github.com/onitake/kanjirecog) Android app.

## Demo

[Jiten Japanese Dictionary](https://jiten.obfusk.dev)
uses `kanjidraw` with a
[JavaScript frontend](https://github.com/obfusk/jiten/blob/master/jiten/static/kanjidraw.js).

## Requirements

* Python >= 3.5 (w/ Tk support for the GUI).

### Debian/Ubuntu

```bash
$ apt install python3-tk
```

## Installing

### Using pip

```bash
$ pip install kanjidraw
```

NB: depending on your system you may need to use e.g. `pip3 --user`
instead of just `pip`.

### From git

NB: this installs the latest development version, not the latest
release.

```bash
$ git clone https://github.com/obfusk/kanjidraw.git
$ cd kanjidraw
$ pip install -e .
```

NB: you may need to add e.g. `~/.local/bin` to your `$PATH` in order
to run `kanjidraw`.

To update to the latest development version:

```bash
$ cd kanjidraw
$ git pull --rebase
```

## Examples

### Kanji Input on Linux

#### kanjidraw-paste

Opens `kanjidraw` to select one (`--oneshot`) or multiple
(`--multiple`) kanji, and afterwards pastes the selected kanji in the
active window.  Requires `xclip` and `xdotool`.

```bash
#!/bin/bash
set -e
pid="$( xdotool getactivewindow getwindowpid )"
pids() { xdotool search --classname "$1" getwindowpid %@; }
if pids rxvt | grep -q "^$pid$"; then
  key=ctrl+alt+v
elif ( pids terminal; pids st-256color ) | grep -q "^$pid$"; then
  key=ctrl+shift+v
else
  key=ctrl+v
fi
kanjidraw -s "$@" | tr -d '\n' | xclip -i -selection clipboard
xdotool key --delay 250 "$key"
```

#### i3 config

Keybindings for i3.  Creating custom keybindings for `kanjidraw-paste
--oneshot` and/or `kanjidraw-paste --multiple` should work similarly
with other window managers and desktop environments.

```
for_window [title="Kanji Draw"] floating enable
bindsym $mod+Control+k exec --no-startup-id kanjidraw-paste --oneshot
bindsym $mod+Control+m exec --no-startup-id kanjidraw-paste --multiple
```

## Miscellaneous

### GUI Options

```bash
$ kanjidraw --help
usage: kanjidraw [-h] [-s] [-o | -m] [-d] [--version]

optional arguments:
  -h, --help      show this help message and exit
  -s, --stdout    print kanji to stdout instead of copying to clipboard
  -o, --oneshot   quit after one kanji
  -m, --multiple  queue kanji and copy/print after pressing 'c' or quitting
  -d, --dark      use dark theme
  --version       show program's version number and exit
```

Additional keybindings: `q` to quit, `<esc>` to go back.

### Enabling Dark Mode

```bash
$ export KANJIDRAW_DARK=1
```

### Disabling the Grid

```bash
$ export KANJIDRAW_NOGRID=1
```

## License

### Code

© Felix C. Stegerman

[![AGPLv3+](https://www.gnu.org/graphics/agplv3-155x51.png)](https://www.gnu.org/licenses/agpl-3.0.html)

### KanjiVG (stroke data)

© Ulrich Apel

[![CC-BY-SA](https://licensebuttons.net/l/by-sa/3.0/88x31.png)](https://github.com/KanjiVG/kanjivg/blob/master/COPYING)

<!-- vim: set tw=70 sw=2 sts=2 et fdm=marker : -->


debian/changelog
kanjidraw (0.2.3-1ppa.focal1) focal; urgency=medium

  * Ubuntu focal requires debhelper-compat = 12 (and -b ubuntu/focal).
  * v0.2.3 for Ubuntu focal.

 -- Felix C. Stegerman <flx@obfusk.net>  Thu, 20 May 2021 18:52:24 +0200


kanjidraw.py
#!/usr/bin/python3
# encoding: utf-8

# --                                                            ; {{{1
#
# File        : kanjidraw.py
# Maintainer  : Felix C. Stegerman <flx@obfusk.net>
# Date        : 2021-05-09
#
# Copyright   : Copyright (C) 2021  Felix C. Stegerman
# Version     : v0.0.1
# License     : AGPLv3+
#
# --                                                            ; }}}1

                                                                # {{{1
r"""

...

"""                                                             # }}}1

import gzip, re
import xml.etree.ElementTree as ET

from collections import namedtuple
from enum import Enum

ARGS_RX = re.compile(r"(-?(?:\d*\.)?\d+)[\s,]*")
PATH_RX = re.compile(r"([MZCS])\s*((?:(?:-?(?:\d*\.)?\d+)[\s,]*)*)", re.I)

DIRECTION_THRESHOLD = 51
DIAGONAL_THRESHOLD  = 77

Line = namedtuple("Line", "x1 y1 x2 y2".split())

DirAndLoc = namedtuple("DirAndLoc", "starts ends dirs moves".split())

class Direction(Enum):                                          # {{{1
  X, N, NE, E, SE, S, SW, W, NW = range(-1, 8)

  @classmethod
  def of_line(cls, l, threshold = DIRECTION_THRESHOLD):
    x1, y1, x2, y2 = l
    dx, dy = x2 - x1, y2 - y1
    adx, ady = abs(dx), abs(dy)
    if adx < threshold and ady < threshold:
      return cls.X
    if adx > ady:
      diag = ady > DIAGONAL_THRESHOLD * adx // 256
      if dx > 0:
        return (cls.NE if dy < 0 else cls.SE) if diag else cls.E
      else:
        return (cls.NW if dy < 0 else cls.SW) if diag else cls.W
    else:
      diag = adx > DIAGONAL_THRESHOLD * ady // 256
      if dy > 0:
        return (cls.SW if dx < 0 else cls.SE) if diag else cls.S
      else:
        return (cls.NW if dx < 0 else cls.NE) if diag else cls.N

  @classmethod
  def of_move(cls, l1, l2, threshold = DIRECTION_THRESHOLD):
    return cls.of_line(Line(*(l1[2:] + l2[:2])), threshold)

  def isclose(a, b):
    return (a == cls.X or b == cls.X or a == b) or \
           (a == ((b+1)%8) or ((a+1)%8) == b)
                                                                # }}}1

class Location(Enum):                                           # {{{1
  N   = (1, 0)
  NE  = (2, 0)
  E   = (2, 1)
  SE  = (2, 2)
  S   = (1, 2)
  SW  = (0, 2)
  W   = (0, 1)
  NW  = (0, 0)
  MID = (1, 1)

  @classmethod
  def of_point(cls, x, y):
    if x < 85:
      return cls.NW if y < 85 else (cls.W if y < 170 else cls.SW)
    elif x < 170:
      return cls.N if y < 85 else (cls.MID if y < 170 else cls.S)
    else:
      return cls.NE if y < 85 else (cls.E if y < 170 else cls.SE)

  def isclose(a, b):
    return abs(a[0] - b[0]) <= 1 and abs(a[1] - b[1]) <= 1
                                                                # }}}1

# ... TODO ...

# FIXME: better error messages
def compare_strict(a, b):                                       # {{{1
  if len(a) != len(b):
    raise ValueError("must have same length")                 #  FIXME
  ...
                                                                # }}}1

def _directions_and_locations(lines):
  return DirAndLoc(
    tuple( Location.of_point(*l[:2]) for l in lines ),
    tuple( Location.of_point(*l[2:]) for l in lines ),
    tuple(map(Direction.of_line, lines)),
    tuple(map(Direction.of_move, lines[1:], lines[:-1]))
  )

def parse_kanjivg(file):                                        # {{{1
  data = {}
  with gzip.open(file) as f:
    for e in ET.parse(f).getroot():
      if e.tag != "kanji": continue
      code  = int(e.get("id").replace("kvg:kanji_", ""), 16)
      char  = chr(code)
      paths = tuple( _path_to_line(p.get("d")) for p in e.findall(".//path") )
      if not (0x4e00 <= code <= 0x9fff): continue             #  FIXME
      assert char not in data
      data[char] = paths
  return data
                                                                # }}}1

# FIXME: handle errors; normalise?
# https://www.w3.org/TR/SVG/paths.html
def _path_to_line(path):                                        # {{{1
  assert path[0] in "Mm"                                      #  FIXME
  for i, m in enumerate(PATH_RX.finditer(path)):
    cmd, args = m.group(1), tuple(map(float, ARGS_RX.findall(m.group(2))))
    if cmd in "Mm":   # moveto
      assert i == 0
      x1, y1 = x2, y2 = args
    elif cmd in "Zz": # closepath
      assert len(args) == 0
      x2, y2 = x1, y1
    elif cmd in "Cc": # curveto
      while args:
        _x1, _y1, _x2, _y2, x, y, *args = args
        x2, y2 = (x, y) if cmd.isupper() else (x2 + x, y2 + y)
    elif cmd in "Ss": # shorthand/smooth curveto
      _x2, _y2, x, y = args
      x2, y2 = (x, y) if cmd.isupper() else (x2 + x, y2 + y)
    else:
      assert False
  assert all( 0 <= v < 109 for v in [x1, y1, x2, y2] )
  return Line(*( int(v * 255 / 109) for v in [x1, y1, x2, y2] ))
                                                                # }}}1

if __name__ == "__main__":
  if "--doctest" in sys.argv:
    verbose = "--verbose" in sys.argv
    import doctest
    if doctest.testmod(verbose = verbose)[0]: sys.exit(1)

# vim: set tw=70 sw=2 sts=2 et fdm=marker :


kanjidraw/__main__.py
#!/usr/bin/python3
# encoding: utf-8

# --                                                            ; {{{1
#
# File        : kanjidraw/__main__.py
# Maintainer  : Felix C. Stegerman <flx@obfusk.net>
# Date        : 2021-05-09
#
# Copyright   : Copyright (C) 2021  Felix C. Stegerman
# Version     : v0.0.1
# License     : AGPLv3+
#
# --                                                            ; }}}1

                                                                # {{{1
r"""

...

"""                                                             # }}}1

import gzip, json, os, re, sys
import xml.etree.ElementTree as ET

from collections import namedtuple
from enum import Enum

DATAFILE = os.path.join(os.path.dirname(__file__), "data.json")

ARGS_RX = re.compile(r"(-?(?:\d*\.)?\d+)[\s,]*")
PATH_RX = re.compile(r"([MZCS])\s*((?:(?:-?(?:\d*\.)?\d+)[\s,]*)*)", re.I)

DIRECTION_THRESHOLD     = 51
DIAGONAL_THRESHOLD      = 77

STROKE_DIRECTION_WEIGHT = 1.0
MOVE_DIRECTION_WEIGHT   = 0.8
STROKE_LOCATION_WEIGHT  = 0.6
CLOSE_WEIGHT            = 0.7

Line      = namedtuple("Line", "x1 y1 x2 y2".split())
DirAndLoc = namedtuple("DirAndLoc", "starts ends dirs moves".split())

class Direction(Enum):                                          # {{{1
  X, N, NE, E, SE, S, SW, W, NW = range(-1, 8)

  @classmethod
  def of_line(cls, l, threshold = DIRECTION_THRESHOLD):
    x1, y1, x2, y2 = l
    dx, dy = x2 - x1, y2 - y1
    adx, ady = abs(dx), abs(dy)
    if adx < threshold and ady < threshold:
      return cls.X
    if adx > ady:
      diag = ady > DIAGONAL_THRESHOLD * adx // 256
      if dx > 0:
        return (cls.NE if dy < 0 else cls.SE) if diag else cls.E
      else:
        return (cls.NW if dy < 0 else cls.SW) if diag else cls.W
    else:
      diag = adx > DIAGONAL_THRESHOLD * ady // 256
      if dy > 0:
        return (cls.SW if dx < 0 else cls.SE) if diag else cls.S
      else:
        return (cls.NW if dx < 0 else cls.NE) if diag else cls.N

  @classmethod
  def of_move(cls, l1, l2, threshold = DIRECTION_THRESHOLD):
    return cls.of_line(Line(*(l1[2:] + l2[:2])), threshold)

  def isclose(a, b):
    return (a == a.X or b == a.X or a == b) or \
           (a.value == ((b.value+1)%8) or ((a.value+1)%8) == b.value)
                                                                # }}}1

class Location(Enum):                                           # {{{1
  N   = (1, 0)
  NE  = (2, 0)
  E   = (2, 1)
  SE  = (2, 2)
  S   = (1, 2)
  SW  = (0, 2)
  W   = (0, 1)
  NW  = (0, 0)
  MID = (1, 1)

  @classmethod
  def of_point(cls, x, y):
    if x < 85:
      return cls.NW if y < 85 else (cls.W if y < 170 else cls.SW)
    elif x < 170:
      return cls.N if y < 85 else (cls.MID if y < 170 else cls.S)
    else:
      return cls.NE if y < 85 else (cls.E if y < 170 else cls.SE)

  def isclose(a, b):
    return abs(a.value[0] - b.value[0]) <= 1 and \
           abs(a.value[1] - b.value[1]) <= 1
                                                                # }}}1

# FIXME: better error messages
def strict_match(a, b):                                         # {{{1
  if len(a) != len(b):
    raise ValueError("must have same length")                 #  FIXME
  dal_a = _directions_and_locations(a)
  dal_b = _directions_and_locations(b)
  score, l = 0.0, len(a)
  for i in range(l):
    if dal_a.dirs[i] == dal_b.dirs[i]:
      score += STROKE_DIRECTION_WEIGHT
    elif dal_a.dirs[i].isclose(dal_b.dirs[i]):
      score += STROKE_DIRECTION_WEIGHT * CLOSE_WEIGHT
    if i > 0:
      if dal_a.moves[i-1] == dal_b.moves[i-1]:
        score += MOVE_DIRECTION_WEIGHT
      elif dal_a.moves[i-1].isclose(dal_b.moves[i-1]):
        score += MOVE_DIRECTION_WEIGHT * CLOSE_WEIGHT
    if dal_a.starts[i] == dal_b.starts[i]:
      score += STROKE_LOCATION_WEIGHT
    elif dal_a.starts[i].isclose(dal_b.starts[i]):
      score += STROKE_LOCATION_WEIGHT * CLOSE_WEIGHT
    if dal_a.ends[i] == dal_b.ends[i]:
      score += STROKE_LOCATION_WEIGHT
    elif dal_a.ends[i].isclose(dal_b.ends[i]):
      score += STROKE_LOCATION_WEIGHT * CLOSE_WEIGHT
  m = l * (STROKE_DIRECTION_WEIGHT + 2 * STROKE_LOCATION_WEIGHT) \
    + (l-1) * MOVE_DIRECTION_WEIGHT
  return 100 * score / m
                                                                # }}}1

def _directions_and_locations(lines):
  return DirAndLoc(
    tuple( Location.of_point(*l[:2]) for l in lines ),
    tuple( Location.of_point(*l[2:]) for l in lines ),
    tuple(map(Direction.of_line, lines)),
    tuple(map(Direction.of_move, lines[1:], lines[:-1]))
  )

def parse_kanjivg(file):                                        # {{{1
  data = {}
  with gzip.open(file) as f:
    for e in ET.parse(f).getroot():
      if e.tag != "kanji": continue
      code  = int(e.get("id").replace("kvg:kanji_", ""), 16)
      char  = chr(code)
      paths = tuple( _path_to_line(p.get("d")) for p in e.findall(".//path") )
      if not (0x4e00 <= code <= 0x9fff): continue             #  FIXME
      kanji = data.setdefault(len(paths), {})
      assert char not in kanji
      kanji[char] = paths
  return data
                                                                # }}}1

# FIXME: handle errors; normalise?
# https://www.w3.org/TR/SVG/paths.html
def _path_to_line(path):                                        # {{{1
  assert path[0] in "Mm"                                      #  FIXME
  for i, m in enumerate(PATH_RX.finditer(path)):
    cmd, args = m.group(1), tuple(map(float, ARGS_RX.findall(m.group(2))))
    if cmd in "Mm":   # moveto
      assert i == 0
      x1, y1 = x2, y2 = args
    elif cmd in "Zz": # closepath
      assert len(args) == 0
      x2, y2 = x1, y1
    elif cmd in "Cc": # curveto
      while args:
        _x1, _y1, _x2, _y2, x, y, *args = args
        x2, y2 = (x, y) if cmd.isupper() else (x2 + x, y2 + y)
    elif cmd in "Ss": # shorthand/smooth curveto
      _x2, _y2, x, y = args
      x2, y2 = (x, y) if cmd.isupper() else (x2 + x, y2 + y)
    else:
      assert False
  assert all( 0 <= v < 109 for v in [x1, y1, x2, y2] )
  return Line(*( int(v * 255 / 109) for v in [x1, y1, x2, y2] ))
                                                                # }}}1

def load_json(file = DATAFILE):
  with open(file) as fh:
    return { int(k): v for k, v in json.load(fh).items() }

def save_json(file, data):
  with open(file, "w") as fh:
    json.dump(data, fh, sort_keys = True)

def matches(lines, data, match = strict_match, max_results = 25,
            cutoff = 0.75):
  it = data[len(lines)].items()
  ms = sorted(( (match(lines, l), k) for k, l in it ), reverse = True)
  max_score = ms[0][0]
  for m in ms[:max_results]:
    if m[0] < max_score * cutoff: break
    yield m

if __name__ == "__main__":
  if "--doctest" in sys.argv:
    verbose = "--verbose" in sys.argv
    import doctest
    if doctest.testmod(verbose = verbose)[0]: sys.exit(1)

# vim: set tw=70 sw=2 sts=2 et fdm=marker :


kanjidraw/gui.py
#!/usr/bin/python3
# encoding: utf-8

# --                                                            ; {{{1
#
# File        : kanjidraw/gui.py
# Maintainer  : Felix C. Stegerman <flx@obfusk.net>
# Date        : 2021-05-19
#
# Copyright   : Copyright (C) 2021  Felix C. Stegerman
# Version     : v0.2.3
# License     : AGPLv3+
#
# --                                                            ; }}}1

                                                                # {{{1
r"""

Handwritten kanji recognition: tkinter GUI.

"""                                                             # }}}1

import argparse, os, sys

import tkinter as tk
import tkinter.font

from tkinter import ttk

from .lib import kanji_data, matches, __version__

NAME, TITLE = "kanjidraw", "Kanji Draw"
HEIGHT = WIDTH = 400
COLS, LINEWIDTH, FONTSIZE = 5, 5, 35
FONTS = ("Noto Sans CJK JP", "Noto Sans CJK SC", "Noto Sans CJK TC",
         "IPAexGothic", "IPAGothic")

LIGHT_THEME = dict(
  bg = "#fff", bg_active = "#eee", fg = "#000", btn_fg = "#fff",
  btn_bg = "#375a7f", btn_bg_active = "#1a252f",
  btn_bg_disabled = "#7c93ab", btn_fg_disabled = "#fff",
  canvas_fg = "#000", grid = "#999"
)
DARK_THEME = dict(
  bg = "#333", bg_active = "#222", fg = "#fff", btn_fg = "#fff",
  btn_bg = "#375a7f", btn_bg_active = "#28415b",
  btn_bg_disabled = "#344b63", btn_fg_disabled = "#b7b7b7",
  canvas_fg = "#ddd", grid = "#999"
)

def frame_button(master, size, **kw):
  frame = ttk.Frame(master, width = size, height = size)
  frame.columnconfigure(0, weight = 1)
  frame.rowconfigure(0, weight = 1)
  frame.grid_propagate(False)
  btn = ttk.Button(frame, **kw)
  btn.grid(column = 0, row = 0, sticky = "nsew")
  return frame, btn

def gui(stdout = False, oneshot = False, multiple = False,      # {{{1
        dark = False):
  """Tkinter GUI."""

  nogrid = os.environ.get("KANJIDRAW_NOGRID") in ("1", "true", "yes")
  if os.environ.get("KANJIDRAW_DARK") in ("1", "true", "yes"):
    dark = True

  win = tk.Tk()
  win.title(TITLE)
  win.columnconfigure(0, weight = 1)
  win.rowconfigure(0, weight = 1)

  theme = DARK_THEME if dark else LIGHT_THEME
  fonts = set(tkinter.font.families())
  kanji_font = tk.font.Font()

  for f in FONTS:
    if f in fonts:
      kanji_font.config(family = f)
      break

  kanji_btn_font = kanji_font.copy()
  kanji_btn_font.config(size = FONTSIZE)

  s = ttk.Style()
  s.configure(".", background = theme["bg"], foreground = theme["fg"])
  s.configure("TButton", background = theme["btn_bg"],
                         foreground = theme["btn_fg"])
  s.map(".", background = [("active", theme["bg_active"]),
                           ("disabled", theme["bg"])])
  s.map("TButton", background = [("active", theme["btn_bg_active"]),
                                 ("disabled", theme["btn_bg_disabled"])],
                   foreground = [("active", theme["btn_fg"]),
                                 ("disabled", theme["btn_fg_disabled"])])
  s.configure("Kanji.TButton", font = kanji_btn_font)
  s.configure("Kanji.TLabel", font = kanji_font)

  max_strokes = max(kanji_data().keys())
  drawing, x, y, strokes, lines = False, 0, 0, [], []

  def on_mousedown(event):
    nonlocal drawing, x, y
    if len(strokes) < max_strokes:
      drawing, x, y = True, event.x, event.y
      strokes.append([x * 255.0 / WIDTH, y * 255.0 / HEIGHT])
      lines.append([])
      enable_buttons()

  def on_mousemove(event):
    if drawing: draw_line(event.x, event.y)

  def on_mouseup(event):
    nonlocal drawing
    if drawing:
      draw_line(event.x, event.y)
      drawing = False
      strokes[-1] += [x * 255.0 / WIDTH, y * 255.0 / HEIGHT]
      update_strokes()

  def on_undo():
    if strokes:
      strokes.pop()
      for l in lines.pop(): canvas.delete(l)
      update_strokes()
      if not strokes: disable_buttons()

  def on_clear():
    strokes.clear(); lines.clear()
    canvas.delete("all"); draw_grid()
    update_strokes(); disable_buttons()

  def on_done():
    res_frame = ttk.Frame(win)
    res_btns  = ttk.Frame(res_frame)
    btn_back  = ttk.Button(res_btns, text = "Go Back", command = on_back(res_frame))
    lbl_info  = ttk.Label(res_btns, text = info_txt)
    res_grid  = ttk.Frame(res_frame)

    ms = matches(strokes, fuzzy = var_fuzzy.get(), offby1 = var_ob1.get())
    for i, (_, kanji) in enumerate(ms):
      col, row = i % COLS, i // COLS
      res_grid.columnconfigure(col, weight = 1)
      res_grid.rowconfigure(row, weight = 1)
      frame, btn = frame_button(res_grid, WIDTH // COLS,
                                text = kanji, style = "Kanji.TButton",
                                command = on_select_kanji(res_frame, kanji))
      frame.grid(column = col, row = row, sticky = "nsew")

    btn_back.pack(side = tk.LEFT, padx = 5, pady = 5)
    lbl_info.pack(side = tk.LEFT, padx = 5, pady = 5)
    res_btns.pack(); res_grid.pack()
    res_frame.grid(row = 0, column = 0, sticky = "nsew")
    win.bind("<Escape>", on_back(res_frame))

  def on_back(res_frame):
    def f(event = None):
      win.unbind("<Escape>")
      res_frame.destroy()
    return f

  def on_select_kanji(res_frame, kanji):
    def f():
      nonlocal drawn_kanji
      if multiple:
        drawn_kanji += kanji
        kanji_lbl.config(text = drawn_kanji)
      else:
        (print if stdout else copy_to_clipboard)(kanji)
        if oneshot: win.quit()
      res_frame.destroy()
      on_clear()
    return f

  def draw_line(x2, y2):
    nonlocal x, y
    l = canvas.create_line(x, y, x2, y2, width = LINEWIDTH,
                           capstyle = tk.ROUND, fill = theme["canvas_fg"])
    lines[-1].append(l)
    x, y = x2, y2

  def draw_grid():
    if not nogrid:
      for x in (WIDTH // 3, 2 * WIDTH // 3):
        canvas.create_line(x, 0, x, HEIGHT, fill = theme["grid"])
      for y in (HEIGHT // 3, 2 * HEIGHT // 3):
        canvas.create_line(0, y, WIDTH, y, fill = theme["grid"])

  def disable_buttons():
    for w in [btn_undo, btn_clear, btn_done]: w.config(state = tk.DISABLED)

  def enable_buttons():
    for w in [btn_undo, btn_clear, btn_done]: w.config(state = tk.NORMAL)

  def update_strokes():
    lbl_strokes.config(text = "Strokes: {}".format(len(strokes)))

  def copy_to_clipboard(text):
    win.clipboard_clear()
    win.clipboard_append(text)

  draw_frame  = ttk.Frame(win)
  btns        = ttk.Frame(draw_frame)
  btn_undo    = ttk.Button(btns, text = "Undo", command = on_undo)
  btn_clear   = ttk.Button(btns, text = "Clear", command = on_clear)
  lbl_strokes = ttk.Label(btns, text = "Strokes: 0")
  btn_done    = ttk.Button(btns, text = "Done", command = on_done)
  checks      = ttk.Frame(draw_frame)
  var_fuzzy   = tk.IntVar()
  var_ob1     = tk.IntVar()
  check_fuzzy = ttk.Checkbutton(checks, variable = var_fuzzy,
                                text = "Ignore stroke order & direction")
  check_ob1   = ttk.Checkbutton(checks, variable = var_ob1, text = "± 1 stroke")

  canvas = tk.Canvas(draw_frame, height = HEIGHT, width = WIDTH,
                     background = theme["bg"], highlightthickness = 1,
                     highlightbackground = theme["canvas_fg"])
  canvas.bind("<ButtonPress-1>", on_mousedown)
  canvas.bind("<B1-Motion>", on_mousemove)
  canvas.bind("<ButtonRelease-1>", on_mouseup)

  draw_grid(); disable_buttons()
  for w in [btn_undo, btn_clear, lbl_strokes, btn_done, check_fuzzy, check_ob1]:
    w.pack(side = tk.LEFT, padx = 5, pady = 5)
  btns.pack(); checks.pack(); canvas.pack()
  draw_frame.grid(row = 0, column = 0, sticky = "nsew")

  if multiple:
    info_txt    = "Click to add to queue"                     #  FIXME
    drawn_kanji = ""
    kanji_lbl   = ttk.Label(draw_frame, style = "Kanji.TLabel")
    kanji_lbl.pack()

    def output():
      nonlocal drawn_kanji
      if drawn_kanji:
        (print if stdout else copy_to_clipboard)(drawn_kanji)
        drawn_kanji = ""
        kanji_lbl.config(text = "")

    def quit():
      output()
      win.quit()

    win.bind("c", lambda e: output())
    win.bind("q", lambda e: quit())
    win.protocol("WM_DELETE_WINDOW", quit)
  else:
    if stdout:
      info_txt = "Click to print to stdout"
    else:
      info_txt = "Click to copy to clipboard"
    win.bind("q", lambda e: win.quit())

  win.mainloop()
                                                                # }}}1

def main():                                                     # {{{1
  p = argparse.ArgumentParser(prog = NAME)
  p.add_argument("-s", "--stdout", action = "store_true",
                 help = "print kanji to stdout instead of "
                        "copying to clipboard")
  g = p.add_mutually_exclusive_group()
  g.add_argument("-o", "--oneshot", action = "store_true",
                 help = "quit after one kanji")
  g.add_argument("-m", "--multiple", action = "store_true",
                 help = "queue kanji and copy/print after "
                        "pressing 'c' or quitting")
  p.add_argument("-d", "--dark", action = "store_true",
                 help = "use dark theme")
  p.add_argument("--version", action = "version",
                 version = "%(prog)s {}".format(__version__))
  gui(**vars(p.parse_args()))
                                                                # }}}1

if __name__ == "__main__":
  if "--doctest" in sys.argv:
    verbose = "--verbose" in sys.argv
    import doctest
    if doctest.testmod(verbose = verbose)[0]: sys.exit(1)
  else:
    main()

# vim: set tw=70 sw=2 sts=2 et fdm=marker :
